hnswlib/hnswalg.h

```diff
@@ -656,6 +700,291 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
+        // Generally, this storage scheme is a bit naive, and we could do a lot better in terms of disk access patterns
+        this->output_data_level0_.seekp(0, std::ios::beg);
+        for (const auto& id : elements_to_persist_) {
+            // Write the _data_level0_memory
+            // Each element is stored in a contiguous block of size_data_per_element_
+            // Where each element is the the size of llist, the llist, the data, and the label
+            this->output_data_level0_.seekp(id * size_data_per_element_, this->output_data_level0_.beg);
+            this->output_data_level0_.write(data_level0_memory_ + id * size_data_per_element_, size_data_per_element_);
+        }
+        this->output_data_level0_.flush();
+
+        // Write the dirty lengths
+        this->output_length_.seekp(0, std::ios::beg);
+        for (const auto& id : elements_to_persist_) {
+            // Write the _length_memory
+            this->output_length_.seekp(id * sizeof(float), this->output_length_.beg);
+            writeBinaryPOD(this->output_length_, ((float*)length_memory_)[id]);
+        }
+        this->output_length_.flush();
+
+        // Write the dirty link lists
+        this->output_link_lists_.seekp(0, std::ios::beg);
+        auto dirty_elements_iter = elements_to_persist_.begin();
+        // TODO: don't need to iterate over potentially all elements, could store it or memoize
+        for (size_t i = 0; i < cur_element_count && dirty_elements_iter != elements_to_persist_.end(); i++) {
+            unsigned int linkListSize = element_levels_[i] > 0 ? size_links_per_element_ * element_levels_[i] : 0;
+            if (i == *dirty_elements_iter) {
+                writeBinaryPOD(this->output_link_lists_, linkListSize);
+                if (linkListSize)
+                    this->output_link_lists_.write(linkLists_[i], linkListSize);
+                dirty_elements_iter = std::next(dirty_elements_iter);
+            } else {
+                this->output_link_lists_.seekp(linkListSize + sizeof(unsigned int), this->output_link_lists_.cur);
+            }
+        }
+        this->output_link_lists_.flush();
+
+        // Note: It would make sense to do a fsync here
+        elements_to_persist_.clear();
+    }
+
+    void loadPersistedIndex(SpaceInterface<dist_t> *s, size_t max_elements_i = 0){
+        std::ifstream input_header(this->getHeaderLocation(), std::ios::binary);
+        if (!input_header.is_open())
+            throw std::runtime_error("Cannot open header file");
+
+        // Read the header
+        int persisted_version;
+        readBinaryPOD(input_header, persisted_version);
+        // For now, version is a simple equality check, we may add backwards compatibility later
+        if (persisted_version != PERSISTENCE_VERSION)
+            throw std::runtime_error("Cannot read persisted index: wrong persistence version");
```

Comment:
output the persistence version

Reply:
using std::to_string() is breaking windows due to what seems like a compiler that doesn't support c++11. I don't want to introduce an extra variable so we can revisit this in the future.